index.html

```diff
@@ -1755,6 +1854,96 @@ <h2>
+        handler URL, the user agent SHOULD run <a>Handling a protocol
+        launch</a>.
+      </p>
+      <p class="note">
+        [[HTML]]'s <code>registerProtocolHandler</code> method allows web sites
+        to register themselves as possible handlers for particular protocols.
+        What constitutes valid <code>protocol</code> and <code>url</code>
+        values for <a>protocol handler description</a>s is defined in that API. Also
+        note that the [[HTML]] API uses [=url/scheme=]
+        where we use <code>protocol</code> but the same restrictions apply.
+      </p>
+      <section>
+        <h3>
+          `protocol` member
+        </h3>
+        <p>
+          The <dfn>protocol</dfn> member of a <a>protocol handler description</a> is a
+          <a>string</a> that represents the protocol to be handled, such as
+          `mailto` or `web+auth`.
+        </p>
+        <p>
+          The <a>protocol</a> member of a <a>protocol handler description</a> is
+          equivalent to {{NavigatorContentUtils/registerProtocolHandler()}}'s
+          <a href="https://url.spec.whatwg.org/#concept-url-scheme">scheme</a>
+          argument defined in [[HTML]].
+        </p>
+      </section>
+      <section>
+        <h3>
+          `url` member
+        </h3>
+        <p>
+          The <dfn>url</dfn> member of a <a>protocol handler description</a> is the
+          <a>URL</a> [=manifest/within scope=] of the application that opens
+          when the associated protocol is activated.
+        </p>
+        <p>
+          The <a>url</a> member of a <a>protocol handler description</a> is equivalent
+          to {{NavigatorContentUtils/registerProtocolHandler()}}'s
+          <code>url</code> argument defined in [[HTML]].
+        </p>
+      </section>
+      <section>
+        <h3>
+          <dfn>Handling a protocol launch</dfn>
+        </h3>
+        <p>
+          When a <a>protocol handler description</a> <var>protocol_handler</var> having
+          [=manifest=] <var>manifest</var> is invoked, it goes through the same
+          steps used to [=invoke a protocol handler=] defined in [=HTML=],
+          where the user agent SHOULD navigate to [=url=] and the appropriate
+          browsing context is set to a new top level browsing context.
+        </p>
+      </section>
+      <section>
+        <h3>
+          Privacy and Security considerations
+        </h3>
+        <p>
+          Privacy concerns for custom scheme handling are detailed in the
+          <a data-cite="HTML/system-state#security-and-privacy">
+          Security and privacy section</a> of
+          {{NavigatorContentUtils/registerProtocolHandler()}}.
+        </p>
+        <p>
+          The user agent MUST ask for permission when using a protocol
+          handler for the first time. This feature requires user interaction
+          and a script cannot communicate with another application on its
```

Comment:
This seems odd (not the asking for permission part - but this doesn't feel like it's wired up properly)...  

Using HTML+JS as an example, can you show me what would this look like in practice? If it's to do with "permissions", then I suspect we need some kind of Permissions Policy or Permissions API thing here.

Reply:
Hola Marcos, so the way this would work is that these registrations happen upon installation of the Web App. There is a prompt, and it is similar to the one that appears when registration is invoked from the `.registerProtocolHandler()`. In this case, where the registration happens from the manifest, the prompt appears the first time the protocol handler is invoked for a Web app. (It was a requirement from security/privacy folks).

![image](https://user-images.githubusercontent.com/73939538/119394569-3bf10900-bcca-11eb-966d-4411ea04c7a6.png)

Reply:
>  In this case, where the registration happens from the manifest, the prompt appears the first time the protocol handler is invoked for a Web app.

Sorry, I'm still having a hard tome following. Can you walk me through the set up end-to-end? 

How do I "invoke a protocol handler"? Do I click on a link on some website (appears to be the case from the HTML spec)? 

I'm also confused about this in the HTML spec:

```JS
navigator.registerProtocolHandler('web+soup', 'soup?url=%s')
```

If I can register protocol handlers to do the same thing in HTML using JS, why do we need them in the manifest again? Does this just duplicate the functionality? Or does this work differently or provide some additional functionality that I'm not getting?

Sorry for the dumb questions. I'm just trying to understand.

Reply:
Hola again Marcos! Not sure which dumb questions you are referring to, they all seem important to be clear for me!

The difference between `navigator.registerProtocolHandler()` and having a `protocol_handlers` member in the manifest file is is mostly where/when is the scheme registration and invocation happening. Let me explain a bit further:

- `navigator.registerProtocolHandler()`: once it is called it will immediately try to register the scheme and show the prompt displayed above. This is handy if I want to register a scheme anytime from HTML/JS. Handy AF.
- `protocol_handlers` member in the manifest file: this will register a/several scheme(s) when a Web application is being installed. (register the installed Web app as a possible handler for that scheme). Here a noteworthy difference is that the prompt itself would appear once a user in a website on the browser invokes one of the registered schemes. For example, I installed the BatmanPWA and when I am browsing online I click on a 'web+batman://' link _and then_ the prompt appears in the browser asking if the user wants to use that (registered) PWA. This happens the first time a user invokes a specific scheme. 

So in summary, it is a matter of _where_ the registration happens, as it is more convenient for a user to decide whether they want to use an installed Web app when they run into a link that would require it instead of getting N prompts when trying to install a Wen app (where N = number of valid items in the `protocol_handlers` member in the manifest).

Reply:
There is also the automatic unregistration that happens when a user uninstalls a webapp, which is nice :)

Reply:
> protocol_handlers member in the manifest file: this will register a/several scheme(s) when a Web application is being installed.

Ok, but isn't this in violation of the web's security/permission model? I.e., we only really allow things to be request permission at use time, not _en mass_ at install time. 

Consider, we've previously rejected similar proposals whereby one would do:

```
  features: ["geolocation", "camera", "something else"]
```

This feels very similar.  

> Here a noteworthy difference is that the prompt itself would appear once a user in a website on the browser invokes one of the registered schemes. For example, I installed the BatmanPWA and when I am browsing online I click on a 'web+batman://' link and then the prompt appears in the browser asking if the user wants to use that (registered) PWA. This happens the first time a user invokes a specific scheme. Very convenient indeed.

This is concerning though. If I install BatmanPWA, it could do:

```
   protocol_handlers: [
       /// ...1000 protocols here... MUAHAHAH! 
       /// Appear to handle everything, and show up everywhere!
    ]
```

@dmurph: 
> There is also the automatic unregistration that happens when a user uninstalls a webapp, which is nice :)

Agree, that's nice. But it does presume some tight integration at the OS level with the browser. IIRC (and this was long ago), Firefox couldn't support such things because it could only put a shortcut icon on the home screen, but it had no way of getting notified if a user had deleted the shortcut.

Reply:
Hola Marcos,

> > protocol_handlers member in the manifest file: this will register a/several scheme(s) when a Web application is being installed.
> 
> Ok, but isn't this in violation of the web's security/permission model? I.e., we only really allow things to be request permission at use time, not _en mass_ at install time.
> 

It is as you are saying, the permission request happens when the user is using the protocol for the first time. We think the consent dialog on this first use is a good compromise between asking for permission for every protocol that an app wants to register and the security concerns of registering the protocols.
 
> Consider, we've previously rejected similar proposals whereby one would do:
> 
> ```
>   features: ["geolocation", "camera", "something else"]
> ```
> 
> This feels very similar.
> 
> > Here a noteworthy difference is that the prompt itself would appear once a user in a website on the browser invokes one of the registered schemes. For example, I installed the BatmanPWA and when I am browsing online I click on a 'web+batman://' link and then the prompt appears in the browser asking if the user wants to use that (registered) PWA. This happens the first time a user invokes a specific scheme. Very convenient indeed.
> 
> This is concerning though. If I install BatmanPWA, it could do:
> 
> ```
>    protocol_handlers: [
>        /// ...1000 protocols here... MUAHAHAH! 
>        /// Appear to handle everything, and show up everywhere!
>     ]
> ```
> 

The list of safe-listed protocols is rather small and they're very likely to have multiple handlers already defined among browser, native and installed web apps, which makes it unlikely that an app can take over everything. Would setting an arbitrary limit to the number of protocols that can be registered help? 

> @dmurph:
> 
> > There is also the automatic unregistration that happens when a user uninstalls a webapp, which is nice :)
> 
> Agree, that's nice. But it does presume some tight integration at the OS level with the browser. IIRC (and this was long ago), Firefox couldn't support such things because it could only put a shortcut icon on the home screen, but it had no way of getting notified if a user had deleted the shortcut.